cmd/image-builder/main.go

```diff
@@ -1,20 +1,63 @@
 package main
 
 import (
+	"fmt"
 	"os"
+	"reflect"
 
+	"github.com/osbuild/image-builder/internal/logger"
 	"github.com/osbuild/image-builder/internal/server"
 )
 
+// *string means the value is not required
+// string means the value is required and should have a default value
+func LoadConfigFromEnv(intf interface{}) error {
+	t := reflect.TypeOf(intf).Elem()
+	v := reflect.ValueOf(intf).Elem()
+
+	for i := 0; i < v.NumField(); i++ {
+		fieldT := t.Field(i)
+		fieldV := v.Field(i)
+		key, ok := fieldT.Tag.Lookup("env")
+		if !ok {
+			return fmt.Errorf("No env tag in config field")
+		}
+
+		// If no value is defined in the env try the default value
+		confV, ok := os.LookupEnv(key)
+		if !ok {
+			confV, ok = fieldT.Tag.Lookup("default")
+		}
+
+		kind := fieldV.Kind()
+		if ok {
+			switch kind {
+			case reflect.Ptr:
+				fieldV.Set(reflect.ValueOf(&confV))
+			case reflect.String:
+				fieldV.SetString(confV)
+			default:
+				return fmt.Errorf("Unsupported type")
+			}
+		} else if kind == reflect.String {
+			return fmt.Errorf("Undefined non-pointer field without default value %v", fieldT)
```

Comment:
@larskarlitski what do you think enforcing this? Should I just drop the *string vs string altogether?

Reply:
Yeah, good point. I'd drop it and return an error when anything that is not a string is passed.

In fact, I think I'd drop the whole `default` tag. I don't think it adds anything over just setting the default values before passing it to your function:

```go
config = Config{
        ListenAddress: "localhost:8086",
}

err := LoadConfigFromEnv(&config)
```

Again, this is similar to how package `json` handles defaults (i.e., not).

Reply:
This also makes defaults more explicit – they seem quite hidden in the tags...